### PR TITLE
Feature/update snarkyjs docs - Provable and ProvablePure interfaces

### PR DESCRIPTION
**Pull request by yunus433**

### CODE REVIEW

**Review by MartinMinkov — APPROVED** [review]
This looks great to me! Also tagging @mitschabaude and @barriebyron for review :)

File: src/snarky.d.ts [hunk 1]
```diff
@@ -18,31 +18,124 @@ export {
 export { Snarky, Test, JsonGate };
 
 /**
- * `Provable<T>` is the general circuit type interface. Provable interface describes how a type `T` is made up of field elements and auxiliary (non-field element) data.
+ * `Provable<T>` is the general circuit type interface in SnarkyJS. `Provable<T>` interface describes how a type `T` is made up of {@link Field} elements and "auxiliary" (non-provable) data.
  *
  * You will find `Provable<T>` as the required input type in a few places in SnarkyJS. One convenient way to create a `Provable<T>` is using `Struct`.
```
Comment (barriebyron) [inline src/snarky.d.ts:23]:
```suggestion
 * `Provable<T>` is the required input type in a few places in SnarkyJS. One convenient way to create a `Provable<T>` is using `Struct`.
```

File: src/snarky.d.ts [hunk 2]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * A function that takes `value` (optional), an element of type `T`, as argument and returns an array of any type that made up the "auxilary" (non-provable) data of `value`.
+   * 
+   * @param value - the element of type `T` to generate the auxilary data array from, optional. If not provided, an empty array is returned.
+   * 
+   * @return An array of any type describing how this `T` element is made up of "auxilary" (non-provable) data.
+   */
+  toAuxiliary: (value?: T) => any[];
+
+  /**
+   * A function that return an element of type `T` from the given provable and "auxilary" data.
```
Comment (barriebyron) [inline src/snarky.d.ts:47]:
```suggestion
   * A function that returns an element of type `T` from the given provable and "auxilary" data.
```

File: src/snarky.d.ts [hunk 3]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * **Beware**: Calling the statement `x.sizeInFields()` is equivalent to `x.toFields().length`.
+   * 
+   * @return A `number` representing the size of the `T` type in terms of {@link Field} type.
+   */
   sizeInFields(): number;
-  check: (x: T) => void;
+
+  /**
+   * Assert some statements on the chain to check if `value` element is a valid member of type `T`.
+   * This function does not return anything, rather creates any number of assertions on the chain to prove `value` is a valid member of the type `T`.
```
Comment (barriebyron) [inline src/snarky.d.ts:71]:
```suggestion
   * This function does not return anything, rather it creates any number of assertions on the chain to prove `value` is a valid member of the type `T`.
```

File: src/snarky.d.ts [hunk 4]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * @param value - the element of type `T` to put assertions on.
+   */
+  check: (value: T) => void;
 }
+
 /**
- * `ProvablePure<T>` is a special kind of `Provable<T>`, where the auxiliary data is empty. This means the type only consists of field elements,
- * in that sense it is "pure".
- *
+ * `ProvablePure<T>` is a special kind of {@link Provable} interface, where the "auxiliary" (non-provable) data is empty. This means the type only consists of field elements, in that sense it is "pure".
```
Comment (barriebyron) [inline src/snarky.d.ts:81]:
```suggestion
 * `ProvablePure<T>` is a special kind of {@link Provable} interface, where the "auxiliary" (non-provable) data is empty. This means the type consists only of field elements, in that sense it is "pure".
```

File: src/snarky.d.ts [hunk 5]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * As any element of the interface `ProvablePure<T>` includes no "auxilary" data by definition, this function always return an empty array.
+   * 
+   * @param value - the element of type `T` to generate the auxilary data array from, optional. If not provided, an empty array is returned.
+   * 
+   * @return An empty array, as any element of the interface `ProvablePure<T>` includes no "auxilary" data by definition.
+   */
+  toAuxiliary: (value?: T) => any[];
+
+  /**
+   * A function that return an element of type `T` from the given provable data.
```
Comment (barriebyron) [inline src/snarky.d.ts:109]:
```suggestion
   * A function that returns an element of type `T` from the given provable data.
```

**Review by barriebyron — APPROVED** [review]
@yunus433  You are making great strides, thank you so much for your contributions! I suggested just a few minor grammar updates

File: src/snarky.d.ts [hunk 6]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * **Beware**: Calling the statement `x.sizeInFields()` is equivalent to `x.toFields().length`.
+   * 
+   * @return A `number` representing the size of the `T` type in terms of {@link Field} type.
+   */
   sizeInFields(): number;
-  check: (x: T) => void;
+
+  /**
+   * Assert some statements on the chain to check if `value` element is a valid member of type `T`.
+   * This function does not return anything, rather it creates any number of assertions on the chain to prove `value` is a valid member of the type `T`.
```
Comment (mitschabaude) [inline src/snarky.d.ts:71]: I don't like the terminology "on the chain" to refer to constraints added to the proof. The proof is created locally, and not even necessarily sent to a blockchain (zk proofs are a generic tool, they could be used on Mina, but also in other contexts).

How about this:
```suggestion
   * Add assertions to the proof to check if `value` is a valid member of type `T`.
   * This function does not return anything, instead it creates any number of assertions to prove that `value` is a valid member of the type `T`.
```

File: src/snarky.d.ts [hunk 7]
```diff
@@ -18,31 +18,124 @@ export {
...
+ * `Provable<T>` is the required input type in a few places in SnarkyJS. One convenient way to create a `Provable<T>` is using `Struct`.
  * 
  * The properties and methods on the provable type exist in all base SnarkyJS types as well (aka. {@link Field}, {@link Bool}, etc.). In most cases, a zkApp developer does not need these functions to create Dapps.
  */
 declare interface Provable<T> {
-  toFields: (x: T) => Field[];
-  toAuxiliary: (x?: T) => any[];
-  fromFields: (x: Field[], aux: any[]) => T;
+  /**
+   * A function that takes `value`, an element of type `T`, as argument and returns an array of {@link Field} elements that made up the provable data of `value`.
```
Comment (mitschabaude) [inline src/snarky.d.ts:29]:
```suggestion
   * A function that takes `value`, an element of type `T`, as argument and returns an array of {@link Field} elements that make up the provable data of `value`.
```

File: src/snarky.d.ts [hunk 8]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * @param value - the element of type `T` to generate the {@link Field} array from.
+   * 
+   * @return A {@link Field} array describing how this `T` element is made up of {@link Field} elements.
+   */
+  toFields: (value: T) => Field[];
+
+   /**
+   * A function that takes `value` (optional), an element of type `T`, as argument and returns an array of any type that made up the "auxilary" (non-provable) data of `value`.
+   * 
+   * @param value - the element of type `T` to generate the auxilary data array from, optional. If not provided, an empty array is returned.
```
Comment (mitschabaude) [inline src/snarky.d.ts:40]: > If not provided, an empty array is returned.

That's actually not what it does!

```suggestion
   * @param value - the element of type `T` to generate the auxiliary data array from, optional. If not provided, a default value for auxiliary data is returned.
```

File: src/snarky.d.ts [hunk 9]
```diff
@@ -18,31 +18,124 @@ export {
...
+   */
+  fromFields: (fields: Field[], aux: any[]) => T;
+
+  /**
+   * Return the size of the `T` type in terms of {@link Field} type, as {@link Field} is the primitive type.
+   * 
+   * **Warning**: This function returns a `number`, so you cannot use it to prove something on chain. You can use it during debugging or to understand the memory complexity of some type.
+   * 
+   * **Beware**: Calling the statement `x.sizeInFields()` is equivalent to `x.toFields().length`.
+   *
```
Comment (mitschabaude) [inline src/snarky.d.ts:64]: This is not precise, the `Provable<T>` interface consists of static methods, and is needed internally if we _don't_ have an example instance `x`.

I would suggest to delete this sentence!

```suggestion
   * 
```

File: src/snarky.d.ts [hunk 10]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * @return An element of type `T` generated from the given provable data.
+   */
+  fromFields: (fields: Field[]) => T;
+
+  /**
+   * Return the size of the `T` type in terms of {@link Field} type, as {@link Field} is the primitive type.
+   * 
+   * **Warning**: This function returns a `number`, so you cannot use it to prove something on chain. You can use it during debugging or to understand the memory complexity of some type.
+   * 
+   * **Beware**: Calling the statement `x.sizeInFields()` is equivalent to `x.toFields().length`.
```
Comment (mitschabaude) [inline src/snarky.d.ts:124]: delete, see above

File: src/snarky.d.ts [hunk 11]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * 
+   * @return A {@link Field} array describing how this `T` element is made up of {@link Field} elements.
+   */
+  toFields: (value: T) => Field[];
+
+   /**
+   * A function that takes `value` (optional), an element of type `T`, as argument and returns an array of any type that made up the "auxilary" (non-provable) data of `value`.
+   * As any element of the interface `ProvablePure<T>` includes no "auxilary" data by definition, this function always return an empty array.
+   * 
+   * @param value - the element of type `T` to generate the auxilary data array from, optional. If not provided, an empty array is returned.
```
Comment (mitschabaude) [inline src/snarky.d.ts:102]: fix last sentence, see above

File: src/snarky.d.ts [hunk 12]
```diff
@@ -18,31 +18,124 @@ export {
...
+   * **Beware**: Calling the statement `x.sizeInFields()` is equivalent to `x.toFields().length`.
+   * 
+   * @return A `number` representing the size of the `T` type in terms of {@link Field} type.
+   */
   sizeInFields(): number;
-  check: (x: T) => void;
+
+  /**
+   * Assert some statements on the chain to check if `value` element is a valid member of type `T`.
+   * This function does not return anything, rather creates any number of assertions on the chain to prove `value` is a valid member of the type `T`.
```
Comment (mitschabaude) [inline src/snarky.d.ts:132]: change "on chain" terminology, see above

**Review by mitschabaude — COMMENTED** [review]
Thanks @yunus433, this is a nice addition!

I have a few comments below. Also, I found that you misspelled auxiliary as "auxilary" in a few places, could you search for that and fix all of them?

**Pull request merged**